File: android/app/src/main/java/com/flashairsync/core/SyncIndex.kt
package com.flashairsync.core

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.File

/**
 * Manages persistent sync state for deduplication
 */
class SyncIndex(private val context: Context, private val ssid: String) {

    companion object {
        private const val TAG = "SyncIndex"
    }

    private val seenKeys = mutableSetOf<String>()
    private val indexFile: File

    init {
        // Index file path: sync_index_<ssid>.json
        val sanitizedSSID = ssid.replace("/", "_")
        indexFile = File(context.filesDir, "sync_index_$sanitizedSSID.json")

        // Load existing index
        loadFromDisk()
    }

    // MARK: - Public API

    /**
     * Check if an entry has been previously synced
     */
    fun hasSeen(entry: DirectoryEntry): Boolean {
        return seenKeys.contains(entry.dedupeKey)
    }

    /**
     * Mark an entry as synced
     */
    fun markSeen(entry: DirectoryEntry) {
        seenKeys.add(entry.dedupeKey)
    }

    /**
     * Mark multiple entries as synced
     */
    fun markSeen(entries: List<DirectoryEntry>) {
        entries.forEach { seenKeys.add(it.dedupeKey) }
    }

    /**
     * Get count of seen entries
     */
    fun count(): Int = seenKeys.size

    /**
     * Reset the index (clear all entries)
     */
    suspend fun reset() = withContext(Dispatchers.IO) {
        seenKeys.clear()
        if (indexFile.exists()) {
            indexFile.delete()
        }
        Log.i(TAG, "Index reset")
    }

    /**
     * Persist current state to disk
     */
    suspend fun persist() = withContext(Dispatchers.IO) {
        try {
            // Convert set to JSON object
            val jsonObject = JSONObject()
            seenKeys.forEach { key ->
                jsonObject.put(key, true)
            }

            // Atomic write: write to temp file, then rename
            val tempFile = File(context.filesDir, "${indexFile.name}.tmp")
            tempFile.writeText(jsonObject.toString(2))
            tempFile.renameTo(indexFile)

            Log.d(TAG, "Index persisted (${seenKeys.size} entries)")

        } catch (e: Exception) {
            Log.e(TAG, "Failed to persist index: ${e.message}")
        }
    }

    // MARK: - Persistence

    /**
     * Load index from disk
     */
    private fun loadFromDisk() {
        if (!indexFile.exists()) {
            Log.d(TAG, "No existing index found")
            return
        }

        try {
            val jsonText = indexFile.readText()
            val jsonObject = JSONObject(jsonText)

            seenKeys.clear()
            jsonObject.keys().forEach { key ->
                seenKeys.add(key)
            }

            Log.d(TAG, "Loaded index with ${seenKeys.size} entries")

        } catch (e: Exception) {
            Log.e(TAG, "Failed to load index from ${indexFile.path}: ${e.message}")
            seenKeys.clear()
        }
    }

    /**
     * Export index as JSON string (for debugging/support)
     */
    fun exportJSON(): String {
        val jsonObject = JSONObject()
        seenKeys.forEach { key ->
            jsonObject.put(key, true)
        }
        return jsonObject.toString(2)
    }

    /**
     * Get statistics about the index
     */
    fun getStats(): IndexStats {
        return IndexStats(
            totalEntries = seenKeys.size,
            fileURL = indexFile.absolutePath,
            fileSizeBytes = if (indexFile.exists()) indexFile.length() else 0
        )
    }
}

// MARK: - Supporting Types

data class IndexStats(
    val totalEntries: Int,
    val fileURL: String,
    val fileSizeBytes: Long
) {
    val fileSizeFormatted: String
        get() {
            val kb = fileSizeBytes / 1024.0
            val mb = kb / 1024.0
            return when {
                mb >= 1.0 -> String.format("%.2f MB", mb)
                kb >= 1.0 -> String.format("%.2f KB", kb)
                else -> "$fileSizeBytes B"
            }
        }
}

File: ios/FlashAirSync/FlashAirSync/Core/SyncIndex.swift
import Foundation

/// Manages persistent sync state for deduplication
actor SyncIndex {
    private var seenKeys: Set<String>
    private let ssid: String
    private let fileURL: URL

    init(ssid: String) {
        self.ssid = ssid
        self.seenKeys = []

        // Determine storage location
        let docsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let syncDir = docsDir.appendingPathComponent("FlashAirSync", isDirectory: true)

        // Create directory if needed
        try? FileManager.default.createDirectory(at: syncDir, withIntermediateDirectories: true)

        // Index file path: sync_index_<ssid>.json
        let sanitizedSSID = ssid.replacingOccurrences(of: "/", with: "_")
        self.fileURL = syncDir.appendingPathComponent("sync_index_\(sanitizedSSID).json")

        // Load existing index
        self.seenKeys = Self.loadFromDisk(url: fileURL)
    }

    // MARK: - Public API

    /// Check if an entry has been previously synced
    func hasSeen(_ entry: DirectoryEntry) -> Bool {
        seenKeys.contains(entry.dedupeKey)
    }

    /// Mark an entry as synced
    func markSeen(_ entry: DirectoryEntry) {
        seenKeys.insert(entry.dedupeKey)
    }

    /// Mark multiple entries as synced
    func markSeen(_ entries: [DirectoryEntry]) {
        for entry in entries {
            seenKeys.insert(entry.dedupeKey)
        }
    }

    /// Get count of seen entries
    func count() -> Int {
        seenKeys.count
    }

    /// Reset the index (clear all entries)
    func reset() {
        seenKeys.removeAll()
        try? FileManager.default.removeItem(at: fileURL)
    }

    /// Persist current state to disk
    func persist() throws {
        // Convert set to dictionary for JSON encoding
        let dict = Dictionary(uniqueKeysWithValues: seenKeys.map { ($0, true) })

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]

        let data = try encoder.encode(dict)

        // Atomic write: write to temp file, then rename
        let tempURL = fileURL.deletingLastPathComponent()
            .appendingPathComponent(UUID().uuidString)

        try data.write(to: tempURL)
        try FileManager.default.replaceItemAt(fileURL, withItemAt: tempURL)
    }

    // MARK: - Persistence

    /// Load index from disk
    private static func loadFromDisk(url: URL) -> Set<String> {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return []
        }

        do {
            let data = try Data(contentsOf: url)
            let dict = try JSONDecoder().decode([String: Bool].self, from: data)
            return Set(dict.keys)
        } catch {
            print("⚠️ Failed to load sync index from \(url.path): \(error)")
            return []
        }
    }

    /// Export index as JSON string (for debugging/support)
    func exportJSON() throws -> String {
        let dict = Dictionary(uniqueKeysWithValues: seenKeys.map { ($0, true) })
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(dict)
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    /// Get statistics about the index
    func getStats() -> IndexStats {
        IndexStats(
            totalEntries: seenKeys.count,
            fileURL: fileURL.path,
            fileSizeBytes: (try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.size] as? Int) ?? 0
        )
    }
}

// MARK: - Supporting Types

struct IndexStats {
    let totalEntries: Int
    let fileURL: String
    let fileSizeBytes: Int

    var fileSizeFormatted: String {
        ByteCountFormatter.string(fromByteCount: Int64(fileSizeBytes), countStyle: .file)
    }
}
